### ISSUE TITLE
Link.expand encodes the URL for ever expansion

### ISSUE THREAD

**Issue by Bert-R** [spring-projects/spring-hateoas#703]
As suggested for #535, we are calling ``expand`` to get rid of optional template parameters. but that it appears that every iteration of expand does an encoding on the URL. 

Before ``expand``, the URL is this:
```
https://host/users/d3055e11-d01f-4d62-960a-f045dced6b8c?tempPassword=2cgRact2uRElbrtR2c%2B6cewRxGBQrWAR&requestingUserId=d3055e11-d01f-4d62-960a-f045dced6b8c
```

After calling ``expand`` with a map with two parameters, each with a null value, the URL has changed into:
```
https://host/users/d3055e11-d01f-4d62-960a-f045dced6b8c?tempPassword=2cgRact2uRElbrtR2c%25252B6cewRxGBQrWAR&requestingUserId=d3055e11-d01f-4d62-960a-f045dced6b8c
```

So ``%2B`` got encoded to ``%252B`` and then again, into ``%25252B``

This occurs with version 1.5.7

**Comment by odrotbohm** [comment]
We just added a fix for #1127 that touches that area of code. Would you mind checking whether the current snapshots solve your problem as well?

**Comment by Bert-R** [comment]
Thanks for working on this!
I will try to test that, but it is not entirely trivial. We are currently on Spring Hateoas 0.25. I started the upgrade to 1.0, but then ran into #1122. This is fixed now, so I will resume the upgrade to 1.0, based on the current M1 build.

I will let you know the outcome, but it'll take a couple of days.

**Comment by Bert-R** [comment]
How do I convince Spring Boot 2.2.1 to use a snapshot build of your milestone?
I'm using Gradle and added the dependency like this:

```
compile "org.springframework.hateoas:spring-hateoas:1.1.0.BUILD-SNAPSHOT"
```

Then I tried setting the version property like this:
```
ext['spring-hateoas.version'] = '1.1.0.BUILD-SNAPSHOT'
```

But it continues to use the wrong version:
```
|    +--- org.springframework.hateoas:spring-hateoas:1.1.0.BUILD-SNAPSHOT -> 1.0.1.RELEASE
```

**Comment by bclozel** [comment]
I've just created an application on start.spring.io:

```groovy
plugins {
	id 'org.springframework.boot' version '2.2.1.RELEASE'
	id 'io.spring.dependency-management' version '1.0.8.RELEASE'
	id 'java'
}

group = 'com.example'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = '1.8'

ext['spring-hateoas.version'] = '1.1.0.BUILD-SNAPSHOT'

repositories {
	mavenCentral()
	maven { url 'https://repo.spring.io/milestone' }
	maven { url 'https://repo.spring.io/snapshot' }
}

dependencies {
	implementation 'org.springframework.boot:spring-boot-starter-hateoas'
	testImplementation('org.springframework.boot:spring-boot-starter-test') {
		exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
	}
}

test {
	useJUnitPlatform()
}
```

and I'm getting as expected:

```bash
./gradlew dependencies | grep hateoas
\--- org.springframework.boot:spring-boot-starter-hateoas -> 2.2.1.RELEASE
     \--- org.springframework.hateoas:spring-hateoas:1.0.1.RELEASE -> 1.1.0.BUILD-SNAPSHOT
\--- org.springframework.boot:spring-boot-starter-hateoas -> 2.2.1.RELEASE
     \--- org.springframework.hateoas:spring-hateoas:1.0.1.RELEASE -> 1.1.0.BUILD-SNAPSHOT
\--- org.springframework.boot:spring-boot-starter-hateoas (n)
\--- org.springframework.boot:spring-boot-starter-hateoas -> 2.2.1.RELEASE
     \--- org.springframework.hateoas:spring-hateoas:1.0.1.RELEASE -> 1.1.0.BUILD-SNAPSHOT
+--- org.springframework.boot:spring-boot-starter-hateoas -> 2.2.1.RELEASE
|    \--- org.springframework.hateoas:spring-hateoas:1.0.1.RELEASE -> 1.1.0.BUILD-SNAPSHOT
+--- org.springframework.boot:spring-boot-starter-hateoas -> 2.2.1.RELEASE
|    \--- org.springframework.hateoas:spring-hateoas:1.0.1.RELEASE -> 1.1.0.BUILD-SNAPSHOT
```

@Bert-R please provide a sample application we can take a look at

**Comment by Bert-R** [comment]
I found the issue. We have a multiproject build. I should set the version number on the main project. I'm going to migrate the project, test this issue and inform you about the result.

**Comment by Bert-R** [comment]
After finishing the migration to 1.1, I tested the issue and can confirm it is resolved. Thanks!

When do you expect to release 1.1.0? Just to get a feeling on how long we will need to keep the migration branch open.

**Comment by odrotbohm** [comment]
If #1127 fixes your issue, there's also been a backport under #1130. I.e. the fix is also included in the 1.0.x branch. 1.0.3 is currently scheduled for mid January for downstream inclusion in Spring Boot 2.2.3.

**Comment by Bert-R** [comment]
That's good news! Is it possible to do the same for #1122? Otherwise that blocks me from adopting 1.0.x

**Comment by odrotbohm** [comment]
#1122 unfortunately requires API changes which is why I am hesitant to port it back. Let's continue the discussion about options over there.

**Comment by Bert-R** [comment]
#1130 was included in 1.0.2.  I've tested this on 1.0.2 and can confirm it is fixed. 
@odrotbohm Thanks!
